File: CalculatorTest.java
import org.junit.jupiter.api.*;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.io.InputStream;

import static org.junit.jupiter.api.Assertions.*;

public class CalculatorTest {
    private Calculator calculator;
    private final ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
    private final PrintStream originalOut = System.out;
    private InputStream originalIn;
    
    @BeforeEach
    void setUp() {
        // Redirect System.out to capture output
        System.setOut(new PrintStream(outputStream));
        originalIn = System.in;
        
        // Reset static variables before each test
        Calculator.total = 0;
        Calculator.fr = true;
        Calculator.operator = 0;
    }
    
    @AfterEach
    void tearDown() {
        // Restore original System.out and System.in
        System.setOut(originalOut);
        System.setIn(originalIn);
    }
    
    @Test
    void testInitialState() {
        assertEquals(0, Calculator.total, "Total should start at 0");
        assertTrue(Calculator.fr, "First run flag should be true initially");
    }
    
    @Test
    void testAdditionOperation() {
        // Simulate input: number 5, operator 1 (addition), number 3
        String input = "5\n1\n3\n7\n";
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        
        Calculator.main(new String[]{});
        
        String output = outputStream.toString();
        assertTrue(output.contains("Total: 8"), "Should show total after addition");
    }
    
    @Test
    void testSubtractionOperation() {
        // Set initial total to 10, then subtract 3
        Calculator.total = 10;
        Calculator.fr = false;
        
        String input = "2\n3\n7\n"; // operator 2 (subtraction), number 3
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        
        // We need to test the op() method directly
        System.setIn(new ByteArrayInputStream("2\n3\n".getBytes()));
        Calculator.scanner = new Scanner(System.in);
        Calculator.op();
        
        assertEquals(7, Calculator.total, 0.001, "10 - 3 should equal 7");
    }
    
    @Test
    void testMultiplicationOperation() {
        Calculator.total = 5;
        Calculator.fr = false;
        
        String input = "3\n4\n"; // operator 3 (multiplication), number 4
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        Calculator.scanner = new Scanner(System.in);
        
        Calculator.op();
        
        assertEquals(20, Calculator.total, 0.001, "5 * 4 should equal 20");
    }
    
    @Test
    void testDivisionOperation() {
        Calculator.total = 20;
        Calculator.fr = false;
        
        String input = "4\n5\n"; // operator 4 (division), number 5
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        Calculator.scanner = new Scanner(System.in);
        
        Calculator.op();
        
        assertEquals(4, Calculator.total, 0.001, "20 / 5 should equal 4");
    }
    
    @Test
    void testDivisionByZero() {
        Calculator.total = 20;
        Calculator.fr = false;
        
        String input = "4\n0\n"; // operator 4 (division), number 0
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        Calculator.scanner = new Scanner(System.in);
        
        Calculator.op();
        
        String output = outputStream.toString();
        assertTrue(output.contains("no"), "Should display 'no' for division by zero");
        assertEquals(20, Calculator.total, 0.001, "Total should remain unchanged after division by zero");
    }
    
    @Test
    void testModuloOperation() {
        Calculator.total = 10;
        Calculator.fr = false;
        
        String input = "5\n3\n"; // operator 5 (modulo), number 3
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        Calculator.scanner = new Scanner(System.in);
        
        Calculator.op();
        
        assertEquals(1, Calculator.total, 0.001, "10 % 3 should equal 1");
    }
    
    @Test
    void testClearTotalOperation() {
        Calculator.total = 100;
        Calculator.fr = false;
        
        String input = "6\n"; // operator 6 (clear)
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        Calculator.scanner = new Scanner(System.in);
        
        Calculator.op();
        
        assertEquals(0, Calculator.total, "Total should be 0 after clear");
        assertTrue(Calculator.fr, "First run flag should be true after clear");
    }
    
    @Test
    void testExitOperation() {
        Calculator.fr = false;
        
        String input = "7\n"; // operator 7 (exit)
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        Calculator.scanner = new Scanner(System.in);
        
        Calculator.op();
        
        String output = outputStream.toString();
        assertTrue(output.contains("Exiting the calculator"), "Should display exit message");
    }
    
    @Test
    void testInvalidOperator() {
        Calculator.fr = false;
        
        String input = "99\n"; // invalid operator
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        Calculator.scanner = new Scanner(System.in);
        
        Calculator.op();
        
        String output = outputStream.toString();
        assertTrue(output.contains("Invalid operator"), "Should display invalid operator message");
    }
    
    @Test
    void testFirstRunFlow() {
        // Test the complete first run: input number, then addition
        String input = "10\n1\n5\n7\n";
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        
        Calculator.main(new String[]{});
        
        String output = outputStream.toString();
        assertTrue(output.contains("Total: 15"), "First run should add initial number");
        assertTrue(output.contains("Calculator"), "Should display calculator header");
    }
    
    @Test
    void testMultipleOperationsSequence() {
        // Test sequence: 10 + 5 - 3 * 2
        String input = "10\n1\n5\n2\n3\n3\n2\n7\n";
        System.setIn(new ByteArrayInputStream(input.getBytes()));
        
        Calculator.main(new String[]{});
        
        // 10 + 5 = 15, 15 - 3 = 12, 12 * 2 = 24
        String output = outputStream.toString();
        assertTrue(output.contains("Total: 24") || output.contains("Total: 24.0"), 
                  "Should correctly calculate sequence of operations");
    }
}

File: calculator_test.go
package main

import (
	"bytes"
	"strings"
	"testing"

	"calculator/internal/scanner"
)

func TestCalculator_Addition(t *testing.T) {
	input := strings.NewReader("5\n1\n3\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	// Simulate the Run loop manually for testing
	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "Total: 8.00") {
		t.Errorf("Addition failed: expected Total: 8.00, got output:\n%s", output.String())
	}
}

func TestCalculator_Subtraction(t *testing.T) {
	input := strings.NewReader("10\n2\n4\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "Total: 6.00") {
		t.Errorf("Subtraction failed: expected Total: 6.00, got output:\n%s", output.String())
	}
}

func TestCalculator_Multiplication(t *testing.T) {
	input := strings.NewReader("2\n3\n5\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "Total: 10.00") {
		t.Errorf("Multiplication failed: expected Total: 10.00, got output:\n%s", output.String())
	}
}

func TestCalculator_Division(t *testing.T) {
	input := strings.NewReader("20\n4\n4\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "Total: 5.00") {
		t.Errorf("Division failed: expected Total: 5.00, got output:\n%s", output.String())
	}
}

func TestCalculator_DivisionByZero(t *testing.T) {
	input := strings.NewReader("10\n4\n0\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "no") {
		t.Errorf("Division by zero failed: expected 'no' in output, got:\n%s", output.String())
	}
}

func TestCalculator_Modulo(t *testing.T) {
	input := strings.NewReader("10\n5\n3\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "Total: 1.00") {
		t.Errorf("Modulo failed: expected Total: 1.00, got output:\n%s", output.String())
	}
}

func TestCalculator_ClearTotal(t *testing.T) {
	input := strings.NewReader("5\n6\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "Total: 0.00") {
		t.Errorf("Clear total failed: expected Total: 0.00, got output:\n%s", output.String())
	}
}

func TestCalculator_InvalidOperator(t *testing.T) {
	input := strings.NewReader("5\n99\n7\n")
	output := &bytes.Buffer{}
	calc := NewCalculator()
	sc := scanner.NewScanner(input, output)

	calc.fr = true
	var operator int
	for {
		output.WriteString("Calculator\n")
		output.WriteString("==========\n")
		output.WriteString("Total: " + formatFloat(calc.total) + "\n")

		if calc.fr {
			output.WriteString("Input number: ")
			a, err := sc.ReadFloat()
			if err != nil {
				output.WriteString("Invalid input\n")
				continue
			}
			calc.total += a
			calc.fr = false
			operator = simulateOp(calc, sc, output)
		} else {
			operator = simulateOp(calc, sc, output)
		}

		if operator == 7 {
			output.WriteString("Exiting the calculator. Goodbye!\n")
			break
		}
	}

	if !strings.Contains(output.String(), "Invalid operator") {
		t.Errorf("Invalid operator handling failed: expected 'Invalid operator' in output, got:\n%s", output.String())
	}
}

// Helper functions
func formatFloat(f float64) string {
	// Simple formatting to match the calculator's output
	return strings.TrimRight(strings.TrimRight(format.Sprintf("%.2f", f), "0"), ".")
}

func simulateOp(c *Calculator, s *scanner.Scanner, output *bytes.Buffer) int {
	output.WriteString("1. Addition\n")
	output.WriteString("2. Subtraction\n")
	output.WriteString("3. Multiplication\n")
	output.WriteString("4. Division\n")
	output.WriteString("5. Modulo\n")
	output.WriteString("6. Clear Total\n")
	output.WriteString("7. Exit\n")
	output.WriteString("Operator [1...7]: ")

	operator, err := s.ReadInt()
	if err != nil {
		output.WriteString("Invalid operator\n")
		return 0
	}

	switch operator {
	case 1:
		output.WriteString("Input number: ")
		a, err := s.ReadFloat()
		if err != nil {
			output.WriteString("Invalid input\n")
			break
		}
		c.total += a
	case 2:
		output.WriteString("Input number: ")
		a, err := s.ReadFloat()
		if err != nil {
			output.WriteString("Invalid input\n")
			break
		}
		c.total -= a
	case 3:
		output.WriteString("Input number: ")
		a, err := s.ReadFloat()
		if err != nil {
			output.WriteString("Invalid input\n")
			break
		}
		c.total *= a
	case 4:
		output.WriteString("Input number: ")
		a, err := s.ReadFloat()
		if err != nil {
			output.WriteString("Invalid input\n")
			break
		}
		if a != 0 {
			c.total /= a
		} else {
			output.WriteString("no\n")
		}
	case 5:
		output.WriteString("Input number: ")
		a, err := s.ReadFloat()
		if err != nil {
			output.WriteString("Invalid input\n")
			break
		}
		c.total = float64(int(c.total) % int(a))
	case 6:
		c.total = 0
		c.fr = true
	case 7:
		return 7
	default:
		output.WriteString("Invalid operator. Please select a valid option.\n")
	}

	output.WriteString("\n")
	return operator
}
